File: _import/core/ai/go/capabilities.go
package ai

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

const (
	CapabilityManifestSchemaVersion = "capability_manifest.v1"

	CapabilityTenantScopeGlobal  = "global"
	CapabilityTenantScopeOrg     = "org"
	CapabilityTenantScopeProject = "project"

	CapabilityModeRead  = "read"
	CapabilityModeWrite = "write"

	CapabilityRiskLow      = "low"
	CapabilityRiskMedium   = "medium"
	CapabilityRiskHigh     = "high"
	CapabilityRiskCritical = "critical"
)

var (
	capabilitySlugPattern    = regexp.MustCompile(`^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$`)
	capabilityToolPattern    = regexp.MustCompile(`^[a-z][a-z0-9_-]*(?:\.[a-z][a-z0-9_-]*)+$`)
	capabilitySemverPattern  = regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+(?:[-+][0-9A-Za-z.-]+)?$`)
	capabilityTenantScopes   = setOf(CapabilityTenantScopeGlobal, CapabilityTenantScopeOrg, CapabilityTenantScopeProject)
	capabilityModes          = setOf(CapabilityModeRead, CapabilityModeWrite)
	capabilityRiskClasses    = setOf(CapabilityRiskLow, CapabilityRiskMedium, CapabilityRiskHigh, CapabilityRiskCritical)
	forbiddenConfigKeyTokens = []string{
		"apikey",
		"basicauth",
		"baseurl",
		"credential",
		"dsn",
		"endpoint",
		"host",
		"password",
		"secret",
		"token",
		"url",
	}
)

// CapabilityManifest describes a versioned product capability package.
type CapabilityManifest struct {
	SchemaVersion string                      `json:"schema_version"`
	ID            string                      `json:"id"`
	Product       string                      `json:"product"`
	Version       string                      `json:"version"`
	TenantScope   string                      `json:"tenant_scope"`
	Name          string                      `json:"name"`
	Description   string                      `json:"description"`
	Agents        []CapabilityAgentDescriptor `json:"agents"`
	Tools         []CapabilityTool            `json:"tools"`
}

// CapabilityAgentDescriptor is the light routing descriptor exposed by a manifest.
type CapabilityAgentDescriptor struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// CapabilityAuthz declares tenant-local role and module requirements.
type CapabilityAuthz struct {
	RequiredRoles   []string `json:"required_roles"`
	RequiredModules []string `json:"required_modules"`
}

// CapabilityExecutor is a logical execution reference. It is not a URL.
type CapabilityExecutor struct {
	ExecutorRef string `json:"executor_ref"`
}

// CapabilityGovernance declares the Nexus Governance metadata for a tool.
type CapabilityGovernance struct {
	RequiresApproval bool   `json:"requires_approval"`
	ActionType     string `json:"action_type,omitempty"`
	TargetSystem   string `json:"target_system,omitempty"`
}

// CapabilityTool describes a tool that a product capability exposes to Companion.
type CapabilityTool struct {
	Name           string         `json:"name"`
	Description    string         `json:"description"`
	Mode           string         `json:"mode"`
	SideEffect     bool           `json:"side_effect"`
	RiskClass      string         `json:"risk_class"`
	InputSchema    map[string]any `json:"input_schema"`
	OutputSchema   map[string]any `json:"output_schema,omitempty"`
	EvidenceFields []string       `json:"evidence_fields"`
	CapabilityAuthz
	CapabilityExecutor
	Governance *CapabilityGovernance `json:"governance,omitempty"`
}

// ToTool converts a capability tool into the runtime LLM tool declaration.
func (t CapabilityTool) ToTool() Tool {
	return Tool{
		Name:        t.Name,
		Description: t.Description,
		Parameters:  t.InputSchema,
	}
}

// Tools converts all manifest tools into runtime LLM tool declarations.
func (m CapabilityManifest) ToolsForLLM() []Tool {
	out := make([]Tool, 0, len(m.Tools))
	for _, tool := range m.Tools {
		out = append(out, tool.ToTool())
	}
	return out
}

// ValidateCapabilityManifest validates the hard rules for capability manifests.
func ValidateCapabilityManifest(m CapabilityManifest) error {
	if strings.TrimSpace(m.SchemaVersion) != CapabilityManifestSchemaVersion {
		return fmt.Errorf("schema_version must be %q", CapabilityManifestSchemaVersion)
	}
	if err := validateCapabilitySlug("id", m.ID); err != nil {
		return err
	}
	if err := validateCapabilitySlug("product", m.Product); err != nil {
		return err
	}
	if !capabilitySemverPattern.MatchString(strings.TrimSpace(m.Version)) {
		return fmt.Errorf("version must be semver without leading v")
	}
	if !capabilityTenantScopes[m.TenantScope] {
		return fmt.Errorf("tenant_scope must be one of global, org, project")
	}
	if strings.TrimSpace(m.Name) == "" {
		return fmt.Errorf("name is required")
	}
	if strings.TrimSpace(m.Description) == "" {
		return fmt.Errorf("description is required")
	}
	if len(m.Agents) == 0 {
		return fmt.Errorf("at least one agent descriptor is required")
	}
	if len(m.Tools) == 0 {
		return fmt.Errorf("at least one tool is required")
	}

	agents := map[string]bool{}
	for _, agent := range m.Agents {
		if err := validateCapabilitySlug("agent.name", agent.Name); err != nil {
			return err
		}
		if strings.TrimSpace(agent.Description) == "" {
			return fmt.Errorf("agent %q description is required", agent.Name)
		}
		if agents[agent.Name] {
			return fmt.Errorf("duplicate agent name %q", agent.Name)
		}
		agents[agent.Name] = true
	}

	toolNames := map[string]bool{}
	for _, tool := range m.Tools {
		if toolNames[tool.Name] {
			return fmt.Errorf("duplicate tool name %q", tool.Name)
		}
		toolNames[tool.Name] = true
		if err := validateCapabilityTool(tool); err != nil {
			return err
		}
	}

	payload, err := json.Marshal(m)
	if err != nil {
		return fmt.Errorf("marshal manifest for config scan: %w", err)
	}
	var decoded any
	if err := json.Unmarshal(payload, &decoded); err != nil {
		return fmt.Errorf("decode manifest for config scan: %w", err)
	}
	if err := rejectForbiddenCapabilityConfig(decoded, "manifest"); err != nil {
		return err
	}

	return nil
}

func validateCapabilityTool(t CapabilityTool) error {
	if !capabilityToolPattern.MatchString(strings.TrimSpace(t.Name)) {
		return fmt.Errorf("tool.name %q must use dot notation", t.Name)
	}
	if strings.TrimSpace(t.Description) == "" {
		return fmt.Errorf("tool %q description is required", t.Name)
	}
	if !capabilityModes[t.Mode] {
		return fmt.Errorf("tool %q mode must be read or write", t.Name)
	}
	if !capabilityRiskClasses[t.RiskClass] {
		return fmt.Errorf("tool %q risk_class must be low, medium, high, or critical", t.Name)
	}
	if err := validateJSONSchemaObject("input_schema", t.InputSchema); err != nil {
		return fmt.Errorf("tool %q %w", t.Name, err)
	}
	if len(t.OutputSchema) > 0 {
		if err := validateJSONSchemaObject("output_schema", t.OutputSchema); err != nil {
			return fmt.Errorf("tool %q %w", t.Name, err)
		}
	}
	if strings.TrimSpace(t.ExecutorRef) == "" {
		return fmt.Errorf("tool %q executor_ref is required", t.Name)
	}
	if err := validateNonEmptyList("required_roles", t.RequiredRoles); err != nil {
		return fmt.Errorf("tool %q %w", t.Name, err)
	}
	if err := validateNonEmptyList("required_modules", t.RequiredModules); err != nil {
		return fmt.Errorf("tool %q %w", t.Name, err)
	}
	if err := validateNonEmptyList("evidence_fields", t.EvidenceFields); err != nil {
		return fmt.Errorf("tool %q %w", t.Name, err)
	}

	switch t.Mode {
	case CapabilityModeRead:
		if t.SideEffect {
			return fmt.Errorf("tool %q mode=read requires side_effect=false", t.Name)
		}
		if t.Governance != nil && t.Governance.RequiresApproval {
			return fmt.Errorf("tool %q read tools must not require review", t.Name)
		}
	case CapabilityModeWrite:
		if !t.SideEffect {
			return fmt.Errorf("tool %q mode=write requires side_effect=true", t.Name)
		}
		if len(t.EvidenceFields) == 0 {
			return fmt.Errorf("tool %q write tools require evidence_fields", t.Name)
		}
		if t.Governance == nil {
			return fmt.Errorf("tool %q write tools require governance", t.Name)
		}
		if !t.Governance.RequiresApproval {
			return fmt.Errorf("tool %q write tools require governance.requires_approval=true", t.Name)
		}
		if strings.TrimSpace(t.Governance.ActionType) == "" {
			return fmt.Errorf("tool %q write tools require governance.action_type", t.Name)
		}
	}

	return nil
}

func validateCapabilitySlug(field, value string) error {
	if !capabilitySlugPattern.MatchString(strings.TrimSpace(value)) {
		return fmt.Errorf("%s %q must be a stable slug", field, value)
	}
	return nil
}

func validateJSONSchemaObject(field string, schema map[string]any) error {
	if len(schema) == 0 {
		return fmt.Errorf("%s is required", field)
	}
	if schemaType, ok := schema["type"]; !ok || schemaType != "object" {
		return fmt.Errorf("%s must be a JSON Schema object with type=object", field)
	}
	return nil
}

func validateNonEmptyList(field string, values []string) error {
	for _, value := range values {
		if strings.TrimSpace(value) == "" {
			return fmt.Errorf("%s cannot contain empty values", field)
		}
	}
	return nil
}

func rejectForbiddenCapabilityConfig(value any, path string) error {
	switch typed := value.(type) {
	case map[string]any:
		for key, child := range typed {
			if isForbiddenCapabilityConfigKey(key) {
				return fmt.Errorf("capability manifest contains forbidden configuration key %q at %s", key, path)
			}
			if err := rejectForbiddenCapabilityConfig(child, path+"."+key); err != nil {
				return err
			}
		}
	case []any:
		for i, child := range typed {
			if err := rejectForbiddenCapabilityConfig(child, fmt.Sprintf("%s[%d]", path, i)); err != nil {
				return err
			}
		}
	case string:
		if looksLikeRuntimeAddress(typed) {
			return fmt.Errorf("capability manifest contains runtime address at %s", path)
		}
	}
	return nil
}

func isForbiddenCapabilityConfigKey(key string) bool {
	normalized := strings.NewReplacer("_", "", "-", "", ".", "", " ", "").Replace(strings.ToLower(strings.TrimSpace(key)))
	for _, token := range forbiddenConfigKeyTokens {
		if strings.Contains(normalized, token) {
			return true
		}
	}
	return false
}

func looksLikeRuntimeAddress(value string) bool {
	normalized := strings.ToLower(strings.TrimSpace(value))
	return strings.Contains(normalized, "://")
}

func setOf(values ...string) map[string]bool {
	out := make(map[string]bool, len(values))
	for _, value := range values {
		out[value] = true
	}
	return out
}

File: _import/core/ai/python/src/runtime/domain/capabilities.py
"""Capability manifest contracts for ecosystem AI."""

from __future__ import annotations

import re
from collections.abc import Mapping
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from runtime.domain.models import ToolDeclaration

CAPABILITY_MANIFEST_SCHEMA_VERSION = "capability_manifest.v1"

TenantScope = Literal["global", "org", "project"]
CapabilityMode = Literal["read", "write"]
RiskClass = Literal["low", "medium", "high", "critical"]

_SLUG_PATTERN = re.compile(r"^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$")
_TOOL_PATTERN = re.compile(r"^[a-z][a-z0-9_-]*(?:\.[a-z][a-z0-9_-]*)+$")
_SEMVER_PATTERN = re.compile(r"^[0-9]+\.[0-9]+\.[0-9]+(?:[-+][0-9A-Za-z.-]+)?$")
_FORBIDDEN_CONFIG_KEY_TOKENS = (
    "apikey",
    "basicauth",
    "baseurl",
    "credential",
    "dsn",
    "endpoint",
    "host",
    "password",
    "secret",
    "token",
    "url",
)


class CapabilityAgentDescriptor(BaseModel):
    """Light routing descriptor exposed by a capability manifest."""

    model_config = ConfigDict(extra="forbid")

    name: str
    description: str

    @field_validator("name")
    @classmethod
    def _validate_name(cls, value: str) -> str:
        return _validate_slug("agent.name", value)

    @field_validator("description")
    @classmethod
    def _validate_description(cls, value: str) -> str:
        return _validate_required_text("agent.description", value)


class CapabilityAuthz(BaseModel):
    """Tenant-local role and module requirements."""

    model_config = ConfigDict(extra="forbid")

    required_roles: list[str] = Field(default_factory=list)
    required_modules: list[str] = Field(default_factory=list)

    @field_validator("required_roles", "required_modules")
    @classmethod
    def _validate_string_list(cls, value: list[str]) -> list[str]:
        return _validate_string_list(value)


class CapabilityExecutor(BaseModel):
    """Logical execution reference. This is not a URL."""

    model_config = ConfigDict(extra="forbid")

    executor_ref: str

    @field_validator("executor_ref")
    @classmethod
    def _validate_executor_ref(cls, value: str) -> str:
        return _validate_required_text("executor_ref", value)


class CapabilityGovernance(BaseModel):
    """Nexus Governance metadata for a capability tool."""

    model_config = ConfigDict(extra="forbid")

    requires_approval: bool
    action_type: str | None = None
    target_system: str | None = None

    @field_validator("action_type", "target_system")
    @classmethod
    def _validate_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return _validate_required_text("governance field", value)


class CapabilityTool(BaseModel):
    """Tool contract exposed by a product capability."""

    model_config = ConfigDict(extra="forbid")

    name: str
    description: str
    mode: CapabilityMode
    side_effect: bool
    risk_class: RiskClass
    input_schema: dict[str, Any]
    output_schema: dict[str, Any] | None = None
    required_roles: list[str] = Field(default_factory=list)
    required_modules: list[str] = Field(default_factory=list)
    evidence_fields: list[str] = Field(default_factory=list)
    executor_ref: str
    governance: CapabilityGovernance | None = None

    @field_validator("name")
    @classmethod
    def _validate_tool_name(cls, value: str) -> str:
        value = value.strip()
        if not _TOOL_PATTERN.fullmatch(value):
            raise ValueError("tool.name must use dot notation")
        return value

    @field_validator("description", "executor_ref")
    @classmethod
    def _validate_text(cls, value: str) -> str:
        return _validate_required_text("tool field", value)

    @field_validator("input_schema", "output_schema")
    @classmethod
    def _validate_schema_object(cls, value: dict[str, Any] | None) -> dict[str, Any] | None:
        if value is None:
            return None
        if value.get("type") != "object":
            raise ValueError("JSON Schema must be an object with type=object")
        return value

    @field_validator("required_roles", "required_modules", "evidence_fields")
    @classmethod
    def _validate_lists(cls, value: list[str]) -> list[str]:
        return _validate_string_list(value)

    @model_validator(mode="after")
    def _validate_mode_rules(self) -> Self:
        if self.mode == "read":
            if self.side_effect:
                raise ValueError("mode=read requires side_effect=false")
            if self.governance is not None and self.governance.requires_approval:
                raise ValueError("read tools must not require review")
        if self.mode == "write":
            if not self.side_effect:
                raise ValueError("mode=write requires side_effect=true")
            if not self.evidence_fields:
                raise ValueError("write tools require evidence_fields")
            if self.governance is None:
                raise ValueError("write tools require governance")
            if not self.governance.requires_approval:
                raise ValueError("write tools require governance.requires_approval=true")
            if not self.governance.action_type:
                raise ValueError("write tools require governance.action_type")
        return self

    def to_tool_declaration(self) -> ToolDeclaration:
        """Convert this capability tool into the runtime LLM tool declaration."""

        return ToolDeclaration(
            name=self.name,
            description=self.description,
            parameters=dict(self.input_schema),
        )

    def authz(self) -> CapabilityAuthz:
        return CapabilityAuthz(required_roles=list(self.required_roles), required_modules=list(self.required_modules))

    def executor(self) -> CapabilityExecutor:
        return CapabilityExecutor(executor_ref=self.executor_ref)


class CapabilityManifest(BaseModel):
    """Versioned product capability package."""

    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["capability_manifest.v1"]
    id: str
    product: str
    version: str
    tenant_scope: TenantScope
    name: str
    description: str
    agents: list[CapabilityAgentDescriptor]
    tools: list[CapabilityTool]

    @field_validator("id", "product")
    @classmethod
    def _validate_slug_fields(cls, value: str) -> str:
        return _validate_slug("manifest field", value)

    @field_validator("version")
    @classmethod
    def _validate_version(cls, value: str) -> str:
        value = value.strip()
        if not _SEMVER_PATTERN.fullmatch(value):
            raise ValueError("version must be semver without leading v")
        return value

    @field_validator("name", "description")
    @classmethod
    def _validate_text_fields(cls, value: str) -> str:
        return _validate_required_text("manifest field", value)

    @field_validator("agents", "tools")
    @classmethod
    def _validate_non_empty(cls, value: list[Any]) -> list[Any]:
        if not value:
            raise ValueError("list must contain at least one item")
        return value

    @model_validator(mode="after")
    def _validate_hard_rules(self) -> Self:
        agent_names: set[str] = set()
        for agent in self.agents:
            if agent.name in agent_names:
                raise ValueError(f"duplicate agent name {agent.name!r}")
            agent_names.add(agent.name)

        tool_names: set[str] = set()
        for tool in self.tools:
            if tool.name in tool_names:
                raise ValueError(f"duplicate tool name {tool.name!r}")
            tool_names.add(tool.name)

        _reject_forbidden_config(self.model_dump(mode="json", exclude_none=True), "manifest")
        return self

    def tools_for_llm(self) -> list[ToolDeclaration]:
        return [tool.to_tool_declaration() for tool in self.tools]


def validate_capability_manifest(payload: Mapping[str, Any] | CapabilityManifest) -> CapabilityManifest:
    """Validate and return a capability manifest."""

    if isinstance(payload, CapabilityManifest):
        manifest = payload
        _reject_forbidden_config(manifest.model_dump(mode="json", exclude_none=True), "manifest")
        return manifest
    return CapabilityManifest.model_validate(dict(payload))


def _validate_slug(field: str, value: str) -> str:
    value = value.strip()
    if not _SLUG_PATTERN.fullmatch(value):
        raise ValueError(f"{field} must be a stable slug")
    return value


def _validate_required_text(field: str, value: str) -> str:
    value = value.strip()
    if not value:
        raise ValueError(f"{field} is required")
    return value


def _validate_string_list(values: list[str]) -> list[str]:
    for value in values:
        if not value.strip():
            raise ValueError("list values cannot be empty")
    return values


def _reject_forbidden_config(value: Any, path: str) -> None:
    if isinstance(value, Mapping):
        for key, child in value.items():
            if _is_forbidden_config_key(str(key)):
                raise ValueError(f"capability manifest contains forbidden configuration key {key!r} at {path}")
            _reject_forbidden_config(child, f"{path}.{key}")
        return
    if isinstance(value, list):
        for index, child in enumerate(value):
            _reject_forbidden_config(child, f"{path}[{index}]")
        return
    if isinstance(value, str) and _looks_like_runtime_address(value):
        raise ValueError(f"capability manifest contains runtime address at {path}")


def _is_forbidden_config_key(key: str) -> bool:
    normalized = key.strip().lower().replace("_", "").replace("-", "").replace(".", "").replace(" ", "")
    return any(token in normalized for token in _FORBIDDEN_CONFIG_KEY_TOKENS)


def _looks_like_runtime_address(value: str) -> bool:
    return "://" in value.strip().lower()


__all__ = [
    "CAPABILITY_MANIFEST_SCHEMA_VERSION",
    "CapabilityAgentDescriptor",
    "CapabilityAuthz",
    "CapabilityExecutor",
    "CapabilityGovernance",
    "CapabilityManifest",
    "CapabilityMode",
    "CapabilityTool",
    "RiskClass",
    "TenantScope",
    "validate_capability_manifest",
]
